Support setting empty --resolv-conf

austinmoore- opened the pull request:
Adds support for setting the resolv-conf flag on the kubelet through the cluster spec.

If the `spec.kubelet.resolvConf` field is set to an empty string in the cluster spec, the empty string will carry over to the argument like so:
```
--resolv-conf=
```

If the field is not found in the cluster spec, the argument will simply be omitted.

Resolves #2638

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kops/2662)
<!-- Reviewable:end -->

chrislovecnm commented on pkg/flagbuilder/build_flags.go:
```diff
@@ -131,6 +131,12 @@ func BuildFlags(options interface{}) (string, error) {
 				flag = fmt.Sprintf("--%s=%s", flagName, vString)
 			}
 
+		case *string:
```
Is their a unit test for this, this impacts a lot of code

austinmoore- replied:
I'm currently running this code on 1.6.1, and I've tested three cases:
- The field in the cluster spec is exists, but is empty, resulting in `--resolv-conf=`
- The field in the cluster spec exists, and is not empty, resulting in `--resolv-conf=mystring`
- The field in the cluster spec does not exist, resulting in the flag not being built.

I've also added unit tests for these three conditions in `nodeup/pkg/model/kubelet_test.go`, which indirectly test this code. If you think we need more, I can certainly add tests more specific to the `BuildFlags` function. LMK.

Also, while we're talking about this file, I want to get a second pair of eyes on this change here, since I'm still not very familiar with the ins-and-outs of Go, and reflection in Go: https://github.com/kubernetes/kops/pull/2662/files#diff-8f0a190c9dca8ca0e2ed458eb347580fR73
From my tests, this is working fine, and as expected, but I want to make sure that there's not some weird edge case I'm missing.

chrislovecnm replied:
@robertojrojas / @justinsb / @kris-nova thoughts?

robertojrojas replied:
I think adding a unit test in the BuildFlags for this scenario is a good idea. Specially, if the 'if' condition was modified to fix a specific problem.

chrislovecnm replied:
Does the reflection look ok?

chrislovecnm replied:
Austin i would ask that you add some simple tests just for this code. This is a seriously important piece of logic

chrislovecnm commented on pkg/apis/kops/componentconfig.go:
```diff
@@ -235,9 +235,9 @@ type KubeletConfigSpec struct {
-	//ResolverConfig string `json:"resolvConf"`
+	// ResolverConfig is the resolver configuration file used as the basis
+	// for the container DNS resolution configuration."), []
+	ResolverConfig *string `json:"resolvConf" flag:"resolv-conf"`
```
What k8s versions does this work on? I am wondering if we need to check versions on this.

austinmoore- replied:
I've tested this on 1.6.1. I don't think we need to add a check for this. It seems that the kubelet has supported this flag for a while. The earliest mention of resolv-conf that I could find was [here](https://github.com/kubernetes/kubernetes/commit/1162cfd3b10b3b18aaa8b3b476beea9b1738bdb4) which was from version `v1.2.0-alpha.7`.

chrislovecnm commented:
@k8s-bot ok to test

chrislovecnm commented:
Asked a couple of peeps to take a look at the code block - also since you caught my comment about documentation?

austinmoore- commented:
Thanks. I'll add those tests, and the documentation.

austinmoore- commented:
I moved the tests I originally wrote into `buildflags_test.go` because they're more relevant to testing the flagbuilder than the kubelet component. I also added two more test conditions to cover the cases where pointers are not string pointers. Also in the docs, I added a small note about setting empty arguments. Take a look, and let me know what you think @chrislovecnm

justinsb commented on pkg/flagbuilder/build_flags.go:
```diff
@@ -131,6 +131,12 @@ func BuildFlags(options interface{}) (string, error) {
 
+		case *string:
+			if v != nil {
+				vString := fmt.Sprintf("%v", *v)
```
So actually I think _this_ is fine; we are adding a case here, and we return error on unknown cases, so we know this is additive.   I would prefer us to honor flagEmpty, but we can add that when we first need it.

justinsb replied:
We are adding a case here, and we return error on unknown cases, so we know this is additive.

However, because we are no longer flattening `*string` to `string`, the `*string` block needs to be identical to the `string` block.

I think we need to do this through another tag (like `flag-empty`).  Perhaps a tag `flag-include-empty`?  The idea is that even if the flag is empty, we still want to send it.

If we can do that, and the `*string` and `string` blocks are then identical, this then LGTM.

chrislovecnm commented:
@justinsb we good on this PR now? 

@austinmoore- when you have a chance if you can squash the commits, that would be awesome.  If you want to wait till @justinsb gives you a 👍 no problem.

chrislovecnm commented:
@justinsb assigning to you since you are the last with some comments.

austinmoore- commented:
@chrislovecnm Sure thing

austinmoore- commented:
The Linux Foundation check seems stuck. Is there a way to have it resend the report?

stevemcquaid commented:
If I can do anything to help move this along, LMK :D

chrislovecnm commented:
@stevemcquaid waiting on @justinsb for a final review.

justinsb commented on docs/cluster_spec.md:
```diff
@@ -135,6 +135,15 @@ Will result in the flag `--runtime-config=batch/v2alpha1=true,apps/v1alpha1=true
 
 This block contains configurations for `kubelet`.  See https://kubernetes.io/docs/admin/kubelet/
 
+NOTE: Arguments can be set to empty if the field is included in the spec, and an empty string is passed to it.
```
This probably should say "Some arguments"

justinsb commented:
Looks good.  Going to send a followup for some nits so we can get this in!
